### PR TITLE
optimize: data image compare log

### LINKED ISSUE #1463: when branch rollbacking, the data image compare log confused
- [ ] I have searched the [issues](https://github.com/seata/seata/issues) of this repository and believe that this is not a duplicate.

### Ⅰ. Issue Description
When a branch transaction is rollbacking, the logs of the beforeImage and afterImage are confusing and need to be optimized.

image

### Ⅱ. Describe what happened

  If there is an exception, please attach the exception trace:

```
Just paste your stack trace here!
```

### Ⅳ. How to reproduce it (as minimally and precisely as possible)

1. xxx
2. xxx
3. xxx

### Ⅵ. Environment:

- JDK version :
- OS :
- Others:

### PR DESCRIPTION
**Pull request by zjinlei**
### Ⅱ. Does this pull request fix one issue?

fixes #1463

### CODE REVIEW

File: rm-datasource/src/main/java/io/seata/rm/datasource/DataCompareUtils.java [hunk 1]
```diff
@@ -158,29 +151,21 @@ private static boolean compareRows(TableMeta tableMetaData, List<Row> oldRows, L
...
-                    }
-                    return false;
+                    return Result.buildWithParams(false, "compare row failed, rowKey {}, fieldName {}, reason [newField is null]", rowKey, fieldName);
                 }
-                if (!isFieldEquals(oldField, newField)) {
-                    if (LOGGER.isInfoEnabled()) {
-                        LOGGER.info("compare row failed, rowKey {}, reason [Field not equals]", rowKey);
-                    }
-                    return false;
+                Result<Boolean> oldEqualsNewFieldResult = isFieldEquals(oldField, newField);
```
Comment (slievrly) [inline rm-datasource/src/main/java/io/seata/rm/datasource/DataCompareUtils.java]: It looks different from the previous code logic

File: rm-datasource/src/main/java/io/seata/rm/datasource/undo/AbstractUndoExecutor.java [hunk 2]
```diff
@@ -193,23 +196,35 @@ protected boolean dataValidationAndGoOn(Connection conn) throws SQLException {
...
                 return false;
             } else {
+                if (LOGGER.isInfoEnabled()) {
+                    //insert、delete operation
+                    if (StringUtils.isNotBlank(afterEqualsCurrentResult.getErrMsg())) {
+                        LOGGER.info(afterEqualsCurrentResult.getErrMsg(), afterEqualsCurrentResult.getErrMsgParams());
+                    }
+                    //update operation
+                    else if (StringUtils.isNotBlank(beforeEqualsCurrentResult.getErrMsg())) {
+                        LOGGER.info(beforeEqualsCurrentResult.getErrMsg(), beforeEqualsCurrentResult.getErrMsgParams());
```
Comment (slievrly) [inline rm-datasource/src/main/java/io/seata/rm/datasource/undo/AbstractUndoExecutor.java]: reachable code?

Comment (zjinlei) [reply rm-datasource/src/main/java/io/seata/rm/datasource/undo/AbstractUndoExecutor.java]: removed

File: rm-datasource/src/main/java/io/seata/rm/datasource/undo/AbstractUndoExecutor.java [hunk 3]
```diff
@@ -175,16 +176,18 @@ protected void undoPrepare(PreparedStatement undoPST, ArrayList<Field> undoValue
...
+
         TableRecords beforeRecords = sqlUndoLog.getBeforeImage();
         TableRecords afterRecords = sqlUndoLog.getAfterImage();
 
         // Compare current data with before data
         // No need undo if the before data snapshot is equivalent to the after data snapshot.
-        if (DataCompareUtils.isRecordsEquals(beforeRecords, afterRecords)) {
+        Result<Boolean> beforeEqualsAfterResult = DataCompareUtils.isRecordsEquals(beforeRecords, afterRecords);
+        //result can not be null, use equals can let test cases find problems in the first place
+        if (beforeEqualsAfterResult.getResult().equals(true)) {
```
Comment (jsbxyyx) [inline rm-datasource/src/main/java/io/seata/rm/datasource/undo/AbstractUndoExecutor.java]: if (beforeEqualsAfterResult.getResult()) {...}

Comment (jsbxyyx) [inline rm-datasource/src/main/java/io/seata/rm/datasource/undo/AbstractUndoExecutor.java]: if (beforeEqualsAfterResult.getResult()) {...}

File: rm-datasource/src/main/java/io/seata/rm/datasource/undo/AbstractUndoExecutor.java [hunk 4]
```diff
@@ -193,23 +196,30 @@ protected boolean dataValidationAndGoOn(Connection conn) throws SQLException {
         // Validate if data is dirty.
         TableRecords currentRecords = queryCurrentRecords(conn);
         // compare with current data and after image.
-        if (!DataCompareUtils.isRecordsEquals(afterRecords, currentRecords)) {
-            
+        Result<Boolean> afterEqualsCurrentResult = DataCompareUtils.isRecordsEquals(afterRecords, currentRecords);
+        if (afterEqualsCurrentResult.getResult().equals(false)) {
```
Comment (jsbxyyx) [inline rm-datasource/src/main/java/io/seata/rm/datasource/undo/AbstractUndoExecutor.java]: if (!afterEqualsCurrentResult.getResult()) {...}

File: rm-datasource/src/main/java/io/seata/rm/datasource/undo/AbstractUndoExecutor.java [hunk 5]
```diff
@@ -193,23 +196,30 @@ protected boolean dataValidationAndGoOn(Connection conn) throws SQLException {
...
+
             // If current data is not equivalent to the after data, then compare the current data with the before 
             // data, too. No need continue to undo if current data is equivalent to the before data snapshot
-            if (DataCompareUtils.isRecordsEquals(beforeRecords, currentRecords)) {
+            Result<Boolean> beforeEqualsCurrentResult = DataCompareUtils.isRecordsEquals(beforeRecords, currentRecords);
+            if (beforeEqualsCurrentResult.getResult().equals(true)) {
```
Comment (jsbxyyx) [inline rm-datasource/src/main/java/io/seata/rm/datasource/undo/AbstractUndoExecutor.java]: if (beforeEqualsCurrentResult.getResult()) {...}

File: rm-datasource/src/main/java/io/seata/rm/datasource/undo/AbstractUndoExecutor.java (hunk 3 above)
Comment (zjinlei) [reply rm-datasource/src/main/java/io/seata/rm/datasource/undo/AbstractUndoExecutor.java]: modified 4 places

**Pull request merged**